src/txrequest.cpp

```diff
@@ -0,0 +1,593 @@
+        // be preferred over, it must immediately follow the newly created _READY.
+        auto it_next = std::next(it);
+        if (it_next == m_index.get<ByTxHash>().end() || it_next->m_txhash != it->m_txhash ||
+            it_next->m_state == State::COMPLETED) {
+            // This is the new best CANDIDATE_READY, and there is no IsSelected() announcement for this txhash
+            // already.
+            Modify<ByTxHash>(it, [](Announcement& ann){ ann.m_state = State::CANDIDATE_BEST; });
+        } else if (it_next->m_state == State::CANDIDATE_BEST) {
+            Priority priority_old = m_computer(*it_next);
+            Priority priority_new = m_computer(*it);
+            if (priority_new > priority_old) {
+                // There is a CANDIDATE_BEST announcement already, but this one is better.
+                Modify<ByTxHash>(it_next, [](Announcement& ann){ ann.m_state = State::CANDIDATE_READY; });
+                Modify<ByTxHash>(it, [](Announcement& ann){ ann.m_state = State::CANDIDATE_BEST; });
+            }
+        }
+    }
+
+    //! Change the state of an announcement to something non-IsSelected(). If it was IsSelected(), the next best
+    //! announcement will be marked CANDIDATE_BEST.
+    void ChangeAndReselect(Iter<ByTxHash> it, State new_state)
+    {
+        assert(new_state == State::COMPLETED || new_state == State::CANDIDATE_DELAYED);
+        assert(it != m_index.get<ByTxHash>().end());
+        if (it->IsSelected() && it != m_index.get<ByTxHash>().begin()) {
+            auto it_prev = std::prev(it);
+            // The next best CANDIDATE_READY, if any, immediately precedes the REQUESTED or CANDIDATE_BEST
+            // announcement in the ByTxHash index.
+            if (it_prev->m_txhash == it->m_txhash && it_prev->m_state == State::CANDIDATE_READY) {
+                // If one such CANDIDATE_READY exists (for this txhash), convert it to CANDIDATE_BEST.
+                Modify<ByTxHash>(it_prev, [](Announcement& ann){ ann.m_state = State::CANDIDATE_BEST; });
+            }
+        }
+        Modify<ByTxHash>(it, [new_state](Announcement& ann){ ann.m_state = new_state; });
+    }
+
+    //! Check if 'it' is the only announcement for a given txhash that isn't COMPLETED.
+    bool IsOnlyNonCompleted(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+        assert(it->m_state != State::COMPLETED); // Not allowed to call this on COMPLETED announcements.
+
+        // This announcement has a predecessor that belongs to the same txhash. Due to ordering, and the
+        // fact that 'it' is not COMPLETED, its predecessor cannot be COMPLETED here.
+        if (it != m_index.get<ByTxHash>().begin() && std::prev(it)->m_txhash == it->m_txhash) return false;
+
+        // This announcement has a successor that belongs to the same txhash, and is not COMPLETED.
+        if (std::next(it) != m_index.get<ByTxHash>().end() && std::next(it)->m_txhash == it->m_txhash &&
```

Comment:
But this will miss the case when a COMPLETED announcement is not the immediate next?
{.... it-1, it, it+1, COMPLETED}.

It seems like it actually works properly because it is called for CANDIDATE_BEST (implicit rule). Perhaps add this assert then?

Reply:
> {.... it-1, it, it+1, COMPLETED}.

In that case `it` is not the only non-COMPLETED Announcement for this txhash, so false is the correct return value.

Reply:
Right, thanks!